### ISSUE TITLE
Feature Request: Add -IfNotExists switch to New-Item

### ISSUE THREAD

**Issue by skissane** [PowerShell/PowerShell#11677]
# Summary of the new feature/enhancement

`New-Item` cmdlet can be used to create an empty file. However, it gives an error if the file already exists. Sometimes, you just want to ensure the file exists – create it if it doesn't, do nothing if it does. Of course, one could do this with an `if` conditional but having just a switch to do it would make things a lot easier.

(This is different from `-Force` – `-Force` replaces the existing file with an empty file, `-IfNotExists` would leave the existing file unmodified.)

Sometimes it is suggested that `New-Item` is equivalent to Unix `touch`. However, `touch` leaves existing files intact, `New-Item` obliterates them. If `New-Item` had an `-IfNotExists` option, it would be closer to `touch`.

# Proposed technical implementation details (optional)

**Comment by iSazonov** [comment]
I'd name the parameter -TouchFile. "Touch" says about expected behavior, "File" says that the parameter is a dynamic parameter for FileSystem provider.

/cc @mklement0

**Comment by jhoneill** [comment]
It's a useful thing to be able to do. 
Currently you can do `if (-not (test-path $path)) {new-item $Path} `so this would make the syntax neater, rather than making impossible things possible. . 
It needs to have the right behaviour for different item types. -TouchFile for files only is OK , but something more generic for directories etc. might be better.   
I commonly find I write 
`md $path -erroraction SilentlyContinue `
really I _should_  use `try {} catch{} `and ignore "exists" errors , but stop for "failed to add" ones.  Some sort of "Ensure-there-is " which errors if it can't make one would make for nicer code 

Making a simple powerShell function which is the equivalent of unix's `touch` on Windows would be the natural next thing to do.

**Comment by iSazonov** [comment]
> It needs to have the right behaviour for different item types. -TouchFile for files only is OK , but something more generic for directories etc. might be better.

Makes sense.
Also I found that touch utility can accept optional timestamp.
If it will be a dynamic parameter for FileSystem provider the parameter could be `-Touch <optional datetime>`.

**Comment by mklement0** [comment]
tl;dr:

* Good idea, but the switch should be named `-NoClobber` for consistency with existing switches with essentially the same purpose - this switch should _not_ modify the time stamp of an existing item.

* I suggest leaving time-stamp-related functionality out of `New-Item`.

---

There are two independent aspects:

* ensuring a file's _existence_

* modifying a file's _time stamps_

Even though the Unix `touch` utility combines the two aspects (in its argument-less form, it ensures the existence of the file _and_ updates the last-write time stamp if the file already exists), I don't think that's appropriate for `New-Item`, given its name and purpose.

The _focus_ of `touch` is time-stamp functionality, whereas `New-Item`'s is _existence_.

`touch` has numerous useful options, and fundamentally the ability to target _multiple_ items, via wildcards; the best we could do with a dynamic `-Touch` parameter is to provide a limited, single-item subset of `touch`'s functionality - and I don't think that's worth it.

Food for thought for a fully featured potential PowerShell implementation of the Unix `touch` utility  is in the `Touch-Item` implementation from [this Stack Overflow answer](https://stackoverflow.com/a/58756360/45375); note that I deliberately chose to go with non-standard verb `Touch`, because I couldn't come up with a succinct yet meaningful name using an approved verb.

---

> `md $path -erroraction SilentlyContinue`

As @skissane notes, `-Force`  _truncates_ the item if it is a _file_, but if it is a _directory_, it leaves an existing one alone (including its time stamps), hence:


```powershell
# Create the directory, unless it already exists.
# Note: on Unix-like platforms 'md' and 'mkdir' refer to the Unix mkdir utility, 
# not New-Item.
New-Item -Type Directory -Force $path
```

This asymmetry of the `-Force` behavior is unfortunate, but were's stuck with it.

Therefore, for directories the new `-NoClobber` switch would have to have the same effect as `-Force`.

**Comment by SeeminglyScience** [comment]
What about adding an `Update-Item` that works like `touch`?

**Comment by mklement0** [comment]
`Update` sounds like a good choice for the verb, but `Item` is too vague, I think: the cmdlet wouldn't be updating the _item itself_, but _a piece of associated metadata_.

Perhaps `Update-ItemTimestamp`? Or just `Update-Timestamp`?

A compliant alias name for it would be `udts` (`ud` being the official alias name for the `Update` verb).

Re the term "Item": It suggests support for all providers, which I don't think we need or perhaps even can aim for; unfortunately, there is no better umbrella term for "file or folder" I can think of.

**Comment by iSazonov** [comment]
Update-Path

**Comment by jhoneill** [comment]
> Perhaps `Update-ItemTimestamp`? Or just `Update-Timestamp`?

That makes most sense. 

Back to the original question, I'm not sure how often we need to create an empty file, one can do 
$null  | Add-Content $path  which does nothing to the file if it exists (doesn't change the date stamp) but creates a zero byte file if there is none. 

> * Good idea, but the switch should be named `-NoClobber` for consistency with existing switches with essentially the same purpose - this switch should _not_ modify the timestamp of an existing item

New-item already seems to have `-NoClobber` by default   `-Force` to overcome that deletes the file and creates a new one. Doing append rather than create seems to be the answer for files but for directories or other containers that doesn't make sense.  

What's best depends on whether the aim is to fix general cases and have a switch which does "Ignore 'already exists' errors" or give something very similar to touch for files.

**Comment by SeeminglyScience** [comment]
> `Update` sounds like a good choice for the verb, but `Item` is too vague, I think: the cmdlet wouldn't be updating the _item itself_, but _a piece of associated metadata_.
> Re the term "Item": It suggests support for all providers, which I don't think we need or perhaps even can aim for; unfortunately, there is no better umbrella term for "file or folder" I can think of.

Well, I think it would be up to the provider to determine what updating an item means.  For example, lets say you have a cache provider (I think UD has one?), update item could refresh the cache expiration.  Registry also has a lastwritetime iirc (I know one is recorded somewhere, can't remember if it's surfaced in the dotnet API).

Think of it like `Invoke-Item`.  Plenty of providers do not implement an `Invoke`, but the cmdlet still makes sense.

**Comment by mklement0** [comment]
> Well, I think it would be up to the provider to determine what updating an item means.

Good point; what I was thinking of is updating an item's _content_, for which we do have the noun _Content_, such as in `Set-Content`.

If we can come up with a generic `Update-Item` implementation that meaningfully works for multiple providers, I won't complain, but I wonder if that would become too generic / hard to discover, cause the term _timestamp_ would then be in the parameter names.

Also, such a cross-provider cmdlet would have to have to implement the `touch`-specific functionality via dynamic, filesystem-provider-specific parameters (last-access vs. last-write timestamp, treatment of symbolic links).

But it sounds like at this point we should create a new issue focused on implementing `touch` functionality in a separate cmdlet, and keep this issue focused on the `-NoClobber` aspect.

**Comment by mklement0** [comment]
> `$null | Add-Content $path` which does nothing to the file if it exists (doesn't change the date stamp)

That's a neat _workaround_, but far from obvious; I do think there is value in implementing _desired-state logic_ in this case (and in general): "Do what is necessary to make sure that this file exists - whether it's already there or whether you have to create it".

However, one case I hadn't considered before: if `-Value` is _also_ specified, `-NoClobber` should cause a statement-terminating error, as the two parameters conflict conceptually, given that `New-Item` - by design - when given a `-Value`  - either truncates and fills or creates and fills, leaving only the specified `-Value` in the file (it shouldn't _append_; for that, we have `Add-Content`, which also creates the file on demand).

As an aside: specifying a `-Value` when creating a _directory_ - which arguably shouldn't be prevented _syntactically_ - is currently quietly ignored.

>  deletes the file and creates a new one. 

Actually, it _truncates the existing file_ instead, which is not the same, because truncating preserves the file creation date, permissions, ownership, ...

You could argue that `-Force` should indeed delete and re-create, but truncating is the current behavior.

As for `-NoClobber` vs. `-Force`:

I hadn't consider that `-Force` has _another_ implication:

It creates _parent directories on demand_, so that you can use `New-Item -Force noSuchSubDirYet/newFile` and `New-Item -Force noSuchSubDirYet/another/newDir`.

I we say that `-NoClobber`  means _only_: "leave an existing item alone, (do not also create parent directories on demand)", then adding `-Force` would be needed to ask for the latter, so that we get the following combinations:

switch(es) →  / item type ↓  | `-NoClobber` | `-Force` | `-NoClobber -Force`
--------- | ------------ | ------- | ---------
File           | if file exists, leave it alone, except fail if `-Value` is also specified; if parent path doesn't exist, fail | (current behavior) create parent path on demand; truncate file, if it exists  | if file exists, leave it alone (except if `-Value` is also specified); create parent path on demand
Directory | if directory exists, leave it alone; if parent path doesn't exist, fail | (current behavior) if directory exists, leave it alone; if parent path doesn't exist, create it  | same as `-Force` alone

Note that a slight inconsistency with respect to parameter name `-NoClobber` is that in the existing uses it usually results in an _error_ when the target file already exists, but, given the commonality of leaving an existing target untouched, I think it's still the right name.

**Comment by mklement0** [comment]
P.S., @iSazonov:

> Update-Path

In PowerShell, a _path_ is an _address_ of sorts that locates a provider _item_, as reflected in parameters `-Path` and `-LiteralPath`, so I don't think that's the way to go; it also has echoes of `$env:PATH`. "FileSystemItem" instead of "Item" would capture "file or directory", but that's too wordy.

**Comment by ThomasNieto** [comment]
I don't think an `Update-Item` cmdlet is needed to update the modified date time. `Set-ItemProperty` cmdlet already can update the `LastWriteTime` property of a file.

```powershell
C:\> Get-ItemProperty C:\temp\abc.psd1 -Name lastwritetime

lastwritetime : 1/6/2020 11:22:09 PM
PSPath        : Microsoft.PowerShell.Core\FileSystem::C:\temp\abc.psd1
PSParentPath  : Microsoft.PowerShell.Core\FileSystem::C:\temp
PSChildName   : abc.psd1
PSDrive       : C
PSProvider    : Microsoft.PowerShell.Core\FileSystem


C:\> Set-ItemProperty C:\temp\abc.psd1 -Name lastwritetime -Value (Get-Date)
C:\> Get-ItemProperty C:\temp\abc.psd1 -Name lastwritetime

lastwritetime : 1/26/2020 2:32:33 PM
PSPath        : Microsoft.PowerShell.Core\FileSystem::C:\temp\abc.psd1
PSParentPath  : Microsoft.PowerShell.Core\FileSystem::C:\temp
PSChildName   : abc.psd1
PSDrive       : C
PSProvider    : Microsoft.PowerShell.Core\FileSystem


C:\> Set-ItemProperty C:\temp\abc.psd1 -Name lastwritetime -Value (Get-Date).AddDays(-10)
C:\> Get-ItemProperty C:\temp\abc.psd1 -Name lastwritetime

lastwritetime : 1/16/2020 2:33:05 PM
PSPath        : Microsoft.PowerShell.Core\FileSystem::C:\temp\abc.psd1
PSParentPath  : Microsoft.PowerShell.Core\FileSystem::C:\temp
PSChildName   : abc.psd1
PSDrive       : C
PSProvider    : Microsoft.PowerShell.Core\FileSystem
```

**Comment by mklement0** [comment]
Excellent point, @ThomasNieto  - I completely forgot about `Set-ItemProperty`, because I rarely use it - but it effectively is the embodiment of what we discussed as an all-providers `Update-Item`.

This gives us most of what `touch` can do, except:

* file creation on demand

* being able to distinguish between modifying the properties of a symlink's _target_ and the _link itself_.

Arguably:

* the former can be covered simply by having to call `New-Item [-NoClobber]` first.

* the latter should be added as a dynamic filesystem-provider parameter to `Set-ItemProperty`, such as `-ModifyLink` (default would be to modify the link's _target_).

Given `Set-ItemProperty`'s support for positional arguments and using its built-in alias `sp`, concise commands are possible in interactive use; e.g., the following sets the last-write date of all `*.txt` back by 1 day:

```powershell
sp *.txt LastWriteTime (Get-Date).AddDays(-1)
```

In short: I'm no longer convinced we need a separate `touch`-like cmdlet.

**Comment by mklement0** [comment]
Let me offer a pragmatic solution, which is easy to implement _as a custom solution_ now, also in Windows PowerShell, using either an available `touch` utility or an emulation of it (with the same basis syntax).

On Unix, the `touch` utility is automatically available, so no action is required - just call it directly.

On Windows, there are two options, which you could place in your `$PROFILE`:

* On machines that have WSL installed, you can wrap a Unix distro's `touch` utility as follows:
  * `function touch { wsl touch $args }`

* On other machines, you can use the function below, which emulates the _core_ functionality of the `touch` utility (no options supported).

```powershell
# Define the `touch` function only if a `touch` utility isn't present (it always is on Unix).
if (-not (Get-Command -Type Application touch -ea Ignore)) {
  function touch {
    <#
    .SYNOPSIS
    Emulates the Unix touch utility.
    
    .DESCRIPTION
    On Windows, emulates the core functionality (only) of 
    the Unix touch utility; examples:

    touch new1.txt new2.txt

      Creates said files as empty files in the current dir., if 
      files by that name don't yet exist yet; otherwise sets their
      LastWriteTime to now. Note the use of spaces (only) to
      separate multiple file names.

    touch *.txt

      Sets the LastWriteTime of all matching files, if any, to now.

    Note:
     * No parameter (names) other than -? are supported; all other
       arguments are interpreted as file names. 
     * If you have WSL installed, you can alternatively call
       `wsl touch ...`, which provides access to all features of the
        Unix utility.            
    #>

    # As a courtesy, interpret -h, --help or the absence of arguments the same as -?
    if (-not $args -or $args[0] -in '-h', '--help') { return touch -? }

    # Get any existing files...
    $item = Get-Item -ea SilentlyContinue -ErrorVariable errs -Path $args
    if ($item) { 
      # ... and update their timestamps.
      Write-Verbose "Setting LastWriteTime timestamps to now for: $item"
      $item.ForEach('LastWriteTime', [datetime]::now)
    }
    # Create any files that couldn't be found.
    $errs | ForEach-Object {
      if ($_.Exception -is [System.Management.Automation.ItemNotFoundException]) {
        Write-Verbose "Creating file: $_.TargetObject"
        $null = New-Item -Type File -Path $_.TargetObject
      }
      else {
        # Unexpected error, pass it through.
        $_ | Write-Error
      }
    }
  }
}
```

**Comment by mukundshah** [comment]
Please add a default alias `touch->New-Item`. 
Doesn't matter how differently they work.
If I'm a Linux user and use someone's Windows PC, then I am not going to waste my 30 minutes looking for the equivalent command. I will just type touch and it should work; No throw some "Command not found" error. Doesn't matter what it does, but it should do something, except throwing "Command not found". And if it ain't working, I'm never touching PowerShell, doesn't matter what platform it is.

**Comment by jhoneill** [comment]
> Please add a default alias `touch->New-Item`.
> Doesn't matter how differently they work.
> If I'm a Linux user and use someone's Windows PC, then I am not going to waste my 30 minutes looking for the equivalent command. I will just type touch and it should work; No throw some "Command not found" error. Doesn't matter what it does, but it should do something, except throwing "Command not found". And if it ain't working, I'm never touching PowerShell, doesn't matter what platform it is.

Any shell you run on windows is going throw `command not found`, touch is a unix/linux executable not found in Windows. 
Touch works in pwsh on linux because the utility is there.  

 What you're saying here is if PowerShell fill in unix commands which aren't part of windows you won't use it where the commands are present. And if doesn't get you any nearer to knowing what commands you do need on windows. 

That said,  there are some unix style aliases added to PowerShell on Windows, and `touch` would be easy to add.   It might be better as a feature request 
in its own right.

**Comment by mukundshah** [comment]
> That said, there are some unix style aliases added to PowerShell on Windows, and `touch` would be easy to add. It might be better as a feature request in its own right.

I don't know know how you interpreted my comment. I, myself, might have sounded rude and stupid.

However, having an alias for `touch`, like other commands, will be a great thing to do.

> doesn't get you any nearer to knowing what commands you do need on windows.
We can flag the default alias, and when they are run for the first time, we can prompt the user with a line saying _Hey, `touch` is alias set for `New-Item`. You can instead use `New-Item` or other aliases if there is any._

**Comment by doctordns** [comment]
My view is to NOT define a built-in alias of Touch. That might well confuse users, especially on Linux. I promise you, if this were done, there would be support issues. 

IF the alias is useful for you, then you can add it to your profile.

**Comment by mukundshah** [comment]
can't we just do that for windows??

**Comment by doctordns** [comment]
To some degree, those aliases were added to PowerShell 1.0 to help with early adoption. I am not sure they were ever designed in light of a Linux implementation of PowerShell.

**Comment by jhoneill** [comment]
> To some degree, those aliases were added to PowerShell 1.0 to help with early adoption. I am not sure they were ever designed in light of a Linux implementation of PowerShell.

I think there is a "helping muscle memory" argument - one can make 3 or 4 groups of aliases

1.  I've been using `type` to get file content since 1980-something and PowerShell hasn't forced me to change and use `gc` and someone who has been using `cat` on unix gets the same support. And it works the way either category of user expects.
2.  Command.com and cmd.exe never had `pwd` but it's not harmful to have `pwd`  aliased to `Get-Location` , although the output _looks different_.  `ls` and `ps`  look different - they're _functionally close enough_  but unix switches will sometimes cause a failure,
3. Something like `kill` isn't quite the same (`kill -9 1234` fails so using unix parameters fails most of the time ) but it seems to pass the _functionally close enough_ test
4.  Aliasing `curl` to `invoke-webrequest` will fail in a lot of use cases it isn't _functionally close enough_ so having been in windows powershell but got pulled in pwsh 6. 

Group 4 are things that do more harm than good. "Hey let's alias `grep` to `select-string`"  goes in that box. What about `ifconfig` to `ipconfig.exe` ? that's in group 2 or 3, but "You do it a different way on windows" is valid. 
We've managed without `touch` for all this time but it's really in group 1 or 2.  It wouldn't take much to create it as a function on windows. 
```
if (test-path $path)  {
Set-ItemProperty -Path $Path -Name  LastWriteTime -Value (get-date)
}
else {new-item path} 
```
it probably needs pipeline support, and only output with `-passthru ` because (IIRC) unix seems to have an invisible touch.

**Comment by mukundshah** [comment]
> To some degree, those aliases were added to PowerShell 1.0 to help with early adoption. I am not sure they were ever designed in light of a Linux implementation of PowerShell.

Don't you think these all commands should be as uniform as possible, irrespective of what platform we're using?

**Comment by jhoneill** [comment]
> > To some degree, those aliases were added to PowerShell 1.0 to help with early adoption. I am not sure they were ever designed in light of a Linux implementation of PowerShell.
> 
> Don't you think these all commands should be as uniform as possible, irrespective of what platform we're using?
> Like we have with copy, cut, paste, save, undo, select all, etc.


Ctrl C X V and Z didn't always have those meanings, [Shift] [Insert], is still Paste, [Ctrl] [Insert] is still Copy and [shift] [delete]  is still paste.  If there is an edit menu [Alt] [E] will always open it and C is always copy,  T cut and P paste.  
BTW X is "cut" because it looks like scissors , V is arrow pointing down for Insert (paste) and if you put an arrow head on a Z shape that looks like "go back"  Which is why we don't have ctrl U for undo and ctrl P is usually print not paste

Except ctrl + c  does something different in a windows and unix shell ...   You'd think, say, Microsoft office (which introduced these in the early 1990s, they weren't OS standards originally)  would use a fairly well understood control F for find, it works in notepad, internet explorer, edge, vs code, Word, Excel but it's _forward_ in Outlook.  [window key] [f] is "feed back" (Why?!) [s] for search. [Esc] is 'get out of here without saving' , except in Visio...  


> What about the CLI commands? Why they can't be **similar**?

Well...  you need to go back to the early designers of command.com and ask why, if they had taken < > and | from unix  so weren't _just_  a port of cp/m why they didn't use / for the path support `ls` and `rm`  as well as `dir` and `del`.  Or why `FIND` is totally different in command.com (and its successor cmd.exe) from unix. etc.     
Someone should have got the people implementing DOS / Windows IP stacks and Unix ones together and "banged heads" until they agreed whether the tool should be IPCONFIG or IFConfig ...   
In 1969 when the working practices for C and its derivatives (including unix) were set, CPU cycles were too precious to flatten case so everything was (and remains) case sensitive. CP/M , DOS and Windows saw the world as upper case and converted lower to upper.    

**Some things are going to be different.** 

**Windows** PowerShell was designed as a replacement for cmd.exe , with long verb-noun command names which no-one really wants to type at the command line - hence `Get-Content` has aliases `content` , `gc`, `type` and `cat`.   People like Bruce Payette and Jason Shirk do occasionally comment here one on how some of those early decisions were arrived but basically it seemed like good idea at the time to have `cat`, `ls`, `ps`  and  `sort`  but no one spoke up for `touch` and others. 

**Cross platform** PowerShell hits a problem: if I say "paste `dir | sort lastwritetime` into PowerShell" it works in any Windows version `sort` is a default alias - for `sort-object`, and it sorts by that property.  But pwsh on linux  doesn't create that alias by default (it can be added) so it runs linux's native `sort` and the command fails.  If it were being designed today we probably wouldn't have sort as an alias trumping Windows' sort.exe. PowerShell _prefers_ working as the user of another shell on that OS would expect rather than imposing "one true way" on every OS, but it's not a strong preference ...  If you're on an OS with a `touch` command you get `touch` whether you're in powershell or not, and if your on an OS without - you don't.  

You're right that PowerShell can help people moving between OSes by bring windows-like commands to linux and linux-like commands to linux.  That's not its focus but where it can do so without huge effort, and without creating new problems (like `sort`) its a smart thing to do. "We've always done it that way" and "No one has asked for it" are always bad excuses for not doing something, but the lack of touch doesn't seem to have caused a lot of pain over the 13 or 14 years we've had PowerShell. If you had started Cmd instead you'd need a touch.exe and really that's the best solution but since Windows has managed this far without it I don't see it happening.
